ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java

```diff
@@ -239,12 +245,66 @@ private OperationController getOperationController() {
           new RouterException("Cannot accept operation because Router is closed", RouterErrorCode.RouterClosed);
       routerMetrics.operationDequeuingRate.mark();
       routerMetrics.onDeleteBlobError(routerException);
-      operationCompleteCallback.completeOperation(futureResult, callback, null, routerException);
+      completeOperation(futureResult, callback, null, routerException);
     }
     return futureResult;
   }
 
   /**
+   * Initiated deletes of the blobIds in the given list of ids via the {@link BackgroundDeleter}
+   * @param idsToDelete the list of blobId strings to delete.
+   */
+  private void initiateBackgroundDeletes(List<String> idsToDelete) {
+    for (String id : idsToDelete) {
+      currentOperationsCount.incrementAndGet();
+      currentBackgroundOperationsCount.incrementAndGet();
+      backgroundDeleter.deleteBlob(id, new FutureResult<Void>(), new Callback<Void>() {
+        @Override
+        public void onCompletion(Void result, Exception exception) {
+          if (exception != null) {
+            logger.error("Background delete operation failed with exception", exception);
+          }
+          currentBackgroundOperationsCount.decrementAndGet();
+        }
+      });
+    }
+  }
+
+  /**
+   * Initiate the deletes of the data chunks associated with this blobId, if this blob turns out to be a composite
+   * blob. Note that this causes the rate of gets to increase at the servers.
+   * @param blobId the blobId string associated with the possibly composite blob.
+   */
+  private void initiateChunkDeletesIfAny(final String blobId) {
+    Callback<GetBlobResultInternal> callback = new Callback<GetBlobResultInternal>() {
+      @Override
+      public void onCompletion(GetBlobResultInternal result, Exception exception) {
+        if (exception != null) {
+          logger.error(
+              "Encountered exception when attempting to get chunks of a possibly composite deleted blob" + blobId,
+              exception);
+        } else if (result.getBlobResult != null) {
+          logger.error("Unexpected result returned by background get operation to fetch chunk ids.");
+        } else if (result.storeKeys != null) {
+          // storeKeys can be null if the blob is not a composite blob.
+          List<String> idsToDelete = new ArrayList<>();
```

Comment:
minor: looks like this can be `List<StoreKey>` all the way through (even the `PutManager` has an array of `StoreKey` but converts them to `String`). Keeping it as `List<StoreKey>` and letting `initiateBackgroundDeletes()` do the `toID()` might save you from doing this operation.